napari/layers/utils/color_manager.py

```diff
@@ -0,0 +1,222 @@
+from dataclasses import InitVar
+from typing import Optional, Tuple, Union
+
+import numpy as np
+
+from ...utils.colormaps import CategoricalColormap, Colormap, ensure_colormap
+from ...utils.colormaps.color_transformations import (
+    ColorType,
+    normalize_and_broadcast_colors,
+    transform_color_with_defaults,
+)
+from ...utils.colormaps.standardize_color import transform_single_color
+from ...utils.events.dataclass import Property, evented_dataclass
+from ._color_manager_constants import ColorMode
+from ._color_manager_utils import is_color_mapped
+from .layer_utils import guess_continuous, map_property
+
+
+@evented_dataclass(events=True, properties=True)
+class ColorManager:
+    """Colors for a display property
+
+    Parameters
+    ----------
+    values : np.ndarray
+        The RGBA color for each data entry
+    mode : ColorMode
+        Color setting mode. Should be one of the following:
+        DIRECT (default mode) allows each point to be set arbitrarily
+
+        CYCLE allows the color to be set via a color cycle over an attribute
+
+        COLORMAP allows color to be set via a color map over an attribute
+
+    """
+
+    colors: InitVar[ColorType] = 'black'
```

Comment:
The error comes from using `Optional[np.ndarray]` which causes that type of detection in https://github.com/napari/napari/blob/28e67fe6e39fdbf57a34f0d6751153a1288bf00d/napari/utils/events/dataclass.py#L228-L240 does not detect proper function to be used as compare. 

A simple workaround could be:

```suggestion
    """
    
    __equality_checks__ = {"values": np.array_equal, "current_color": np.array_equal}

    colors: InitVar[ColorType] = 'black'
```

More complex could be an inspection of `_GenericAlias` type annotation which may produce additional work in maintenance.

Reply:
actually... I'm also just realizing that the `_type_to_compare` function from #1966 doesn't work with `InitVars`

Reply:
Ah, that's neat. Thanks, @Czaki !

Reply:
if `annotation = dataclasses.InitVar[str]`, then the actual type is at `annotation.type` ... but `_type_to_compare` will throw a warning.   For instance, this throws a warning:
```python
from napari.utils.events.dataclass import evented_dataclass
from dataclasses import InitVar

@evented_dataclass
class T:
    colors: InitVar[str] = 'black'
```

@Czaki can you submit a PR to fix that?

Reply:
@kevinyamauchi So wait for #2023 merge. Then no fix will be needed.

Reply:
Will do. Thanks for the quick fix, @Czaki !

Reply:
#2023 is now merged